docs/strings.md

```diff
@@ -0,0 +1,464 @@
+ */
+typedef StringEncoding = StringEncodingASCII |
+                         StringEncodingUTF8 |
+                         StringEncodingUTF16 |
+                         StringEncodingUTF32 |
+                         StringEncodingCustom;
+
+/**
+ * Character type, stores a unicode code point.
+ */
+typedef Unichar = U32;
+
+// Placeholder until we have syntax for value types.
+value Range
+{
+	U64 index;
+	U64 length;
+}
+
+/**
+ * Interface for string types.  Defines only the primitive operations that all
+ * strings must implement.
+ */
+interface String
+{
+	/**
+	 * Return the length of the string in Unicode code points.
+	 */
+	func length(self: readonly & String) : size_t;
+
+	/**
+	 * The encoding that this string can most easily convert to.  This will
+	 * typically be the encoding that the string uses internally.
+	 * Note that a return value of `StringEncodingASCII` implies
+	 * `StringEncodingUTF8` for all read operations because ASCII is a strict
+	 * superset of UTF-8.
+	 */
+	func fastest_encoding(self: readonly & String) : StringEncoding;
+
+	/**
+	 * The standard encoding that this string can most easily convert to.  This
+	 * will typically be the encoding that the string uses internally but may
+	 * not be for custom encodings.  For all standard-library string types,
+	 * this will return the same value as `fastest_encoding`.
+	 */
+	func fastest_unicode_encoding(self: readonly & String) : UnicodeStringEncoding;
+
+	/**
+	 * The standard encoding that will use the smallest amount of space to
+	 * store the characters.  For example, most English strings will return
+	 * `StringEncodingASCII`, French will typically return
+	 * `StringEncodingUTF8`, whereas CJK language strings will likely return
+	 * `StringEncodingUTF16`.
+	 */
+	func smallest_unicode_encoding(self: readonly & String) : UnicodeStringEncoding;
+
+	/**
+	 * Access a specific character.  Returns a unicode code point for the
+	 * specific index in the string.
+	 */
+	func apply(self: readonly & String, index: U64) : Unichar;
+
+	/**
+	 * Access a specific character.
+	 */
+	func apply(self: readonly & String, index: U64) : Unichar;
+
+	/**
+	 * Takes a range in the string (expressed in Unicode code points) and
+	 * optionally a buffer.  Either the desired characters (in the specified
+	 * encoding) are copied into that buffer, or returns an immutable array
+	 * that may point to the internal storage for the string.
+	 *
+	 * If the caller first calls `fastest_unicode_encoding` and uses the result
+	 * as the generic parameter then it will return an internal buffer for
+	 * immutable strings.  This is intended to be used to build fast iterators.
+	 */
+	func copy_or_view_data[StringEncoding Enc](self: readonly & String,
+	                                           range: Range,
+	                                           buffer: Optional[Array[Enc.CharType] & mut])
+	     : Array[Enc.CharType] & (mut | imm);
```

Comment:
The use of a path dependent type here is interesting. This seems like an especially interesting function to look at from a type system point of view, to get the best specialisation we can based on the actual type of `self`. It feels like the `buffer` being optional means it should be a separate method if you want to provide it?

Ill-thought out attempt:

```
view[S, Enc](self: S, range: Range & readonly): S~>(Slice[Enc.CharType] & mut)
  where S: String & readonly
  where Enc: StringEncoding
```

My vague idea there is that the Slice (placeholder name for a a view onto an Array) is `mut` if self is `mut` (that is, the Slice can then be used to change the underlying data) and `imm` if self is `imm`. Not sure if that either works or expresses what you want.

Reply:
The idea here, which I don't think I captured well, is to support a fast enumeration pattern where the callee has three choices:

 - Return an internal view of the underlying data (probably only used on immutable instances, might be used on mutable string types that are some complex structure of immutable character runs)
 - Copy data into the buffer provided by the caller.
 - Allocate a new buffer to copy a large chunk into it (if the buffer is not large enough, for example if the caller asks for 128 code points as UTF-8 and provides a 128-byte buffer).

It might be better to make the optional buffer into two different calls.  I don't have a strong opinion either way.  I think your signature captures the variant that doesn't take a buffer correctly.